pkg/plugins/runtime/gateway/traffic_route_generator_test.go

```diff
@@ -0,0 +1,385 @@
+				ModificationTime: core.Now(),
+			})
+			dp.Spec.Networking = &mesh_proto.Dataplane_Networking{
+				Address: fmt.Sprintf("192.168.1.%d", addressCount+1),
+				Inbound: []*mesh_proto.Dataplane_Networking_Inbound{
+					&mesh_proto.Dataplane_Networking_Inbound{
+						Port:        20011,
+						ServicePort: 10011,
+						Tags: map[string]string{
+							mesh_proto.ServiceTag:  serviceName,
+							mesh_proto.ProtocolTag: "http",
+						},
+					}},
+			}
+
+			// Attach tag pairs. It's OK to panic if caller didn't pass even pairs.
+			for i := 0; i < len(tags); i = i + 2 {
+				dp.Spec.GetNetworking().GetInbound()[0].Tags[tags[i]] = tags[i+1]
+			}
+
+			Expect(StoreFixture(rt, dp)).To(Succeed())
+			addressCount++
+		}
+	}
+
+	BeforeEach(func() {
+		var err error
+
+		addressCount = 0
+
+		rt, err = BuildRuntime()
+		Expect(err).To(Succeed(), "build runtime instance")
+
+		Expect(StoreNamedFixture(rt, "mesh-default.yaml")).To(Succeed())
+		Expect(StoreNamedFixture(rt, "dataplane-default.yaml")).To(Succeed())
+		Expect(StoreNamedFixture(rt, "gateway-default.yaml")).To(Succeed())
+	})
+
+	It("should generate from traffic routes", func() {
+		route := `
+type: TrafficRoute
+mesh: default
+name: echo-service
+sources:
+- match:
+    kuma.io/service: gateway-default
+destinations:
+- match:
+    kuma.io/service: '*'
+conf:
+  http:
+  - match:
+      path:
+        prefix: /
+    destination:
+      kuma.io/service: echo-service
+  destination:
+    kuma.io/service: echo-service
+`
+		CreateServiceDataplanes("echo-service", 5)
+
+		Expect(StoreInlineFixture(rt, []byte(route))).To(Succeed())
+		snap, err := Do()
+		Expect(err).To(Succeed())
+
+		out, err := yaml.Marshal(MakeProtoSnapshot(snap))
+		Expect(err).To(Succeed())
+
+		Expect(out).To(matchers.MatchGoldenYAML(path.Join("testdata", "01-traffic-route.yaml")))
+	})
+
+	It("should generate resources from multiple traffic routes", func() {
+		CreateServiceDataplanes("echo-service", 2)
+
+		Expect(StoreInlineFixture(rt, []byte(`
+type: TrafficRoute
+mesh: default
+name: edge-gateway-1
+sources:
+- match:
+    kuma.io/service: gateway-default
+destinations:
+- match:
+    kuma.io/service: '*'
+conf:
+  http:
+  - match:
+      path:
+        prefix: /
+    destination:
+      kuma.io/service: echo-service
+  destination:
+    kuma.io/service: echo-service
+`))).To(Succeed())
+
+		Expect(StoreInlineFixture(rt, []byte(`
+type: TrafficRoute
+mesh: default
+name: edge-gateway-2
+sources:
+- match:
+    kuma.io/service: gateway-default
+destinations:
+- match:
+    kuma.io/service: '*'
+conf:
+  http:
+  - match:
+      path:
+        prefix: /
+    destination:
+      kuma.io/service: echo-service
+  destination:
+    kuma.io/service: echo-service
+`))).To(Succeed())
+
+		snap, err := Do()
+		Expect(err).To(Succeed())
+
+		out, err := yaml.Marshal(MakeProtoSnapshot(snap))
+		Expect(err).To(Succeed())
+
+		Expect(out).To(matchers.MatchGoldenYAML(path.Join("testdata", "02-traffic-route.yaml")))
```

Comment:
If possible I'd like for us to stick to `// given`, `// when`, `// then` convention. At least for me, it greatly improves the readability of the test. I can see what is the "preparation" what is the "action" that we are testing and what is the expected result. Ofc that does not mean that we cannot use Expect in given, because often when we prepare data we need to handle error.

That's my opinion, it would be great to hear from the rest of @kumahq/kuma-maintainers

Reply:
Happy to add anything that makes tests easier to read :)